plugins/dell-dock2/fu-dell-dock2-plugin.c

```diff
@@ -0,0 +1,345 @@
+		locker = fu_device_locker_new(FU_DEVICE(hub_device), error);
+		if (locker == NULL)
+			return FALSE;
+
+		fu_device_add_child(ec_device, FU_DEVICE(hub_device));
+	} else {
+		g_debug("ignoring unknown device, vid: %04X, pid: %04X", vid, pid);
+		return TRUE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_plugin_ec_add_cached_devices(FuPlugin *plugin, FuDevice *ec_device, GError **error)
+{
+	struct {
+		guint16 vid;
+		guint16 pid;
+	} hw_dev_ids[] = {
+	    {DELL_VID, DELL_DOCK2_USB_RTS5480_GEN1_PID},
+	    {DELL_VID, DELL_DOCK2_USB_RTS5480_GEN2_PID},
+	    {DELL_VID, DELL_DOCK2_USB_RTS5485_GEN2_PID},
+	    {0},
+	};
+
+	for (guint i = 0; hw_dev_ids[i].pid != 0; i++) {
+		FuDevice *device;
+		const gchar *key;
+
+		key =
+		    g_strdup_printf("USB\\VID_%04X&PID_%04X", hw_dev_ids[i].vid, hw_dev_ids[i].pid);
+
+		device = fu_plugin_cache_lookup(plugin, key);
+		if (device != NULL) {
+			if (!(fu_dell_dock2_plugin_device_add(plugin, device, error)))
+				return FALSE;
+
+			fu_plugin_cache_remove(plugin, key);
+		}
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_plugin_backend_device_added(FuPlugin *plugin,
+					  FuDevice *device,
+					  FuProgress *progress,
+					  GError **error)
+{
+	g_autofree const gchar *instance_id = NULL;
+	guint16 vid, pid;
+
+	/* not interesting */
+	if (!FU_IS_USB_DEVICE(device))
+		return TRUE;
+
+	/* VID and PID */
+	vid = fu_usb_device_get_vid(FU_USB_DEVICE(device));
+	pid = fu_usb_device_get_pid(FU_USB_DEVICE(device));
+
+	/* USB HUB HID bridge device */
+	if ((vid == DELL_VID && pid == DELL_DOCK2_HID_PID)) {
+		g_autoptr(FuDellDock2Ec) ec_dev = NULL;
+		g_autoptr(GError) error_local = NULL;
+
+		ec_dev = fu_dell_dock2_ec_new(device);
+		if (ec_dev == NULL) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "can't create EC V2 device");
+			return FALSE;
+		}
+
+		if (fu_dell_dock2_plugin_create_node(plugin, FU_DEVICE(ec_dev), &error_local)) {
+			/* flush the cached devices to plugin */
+			if (!fu_dell_dock2_plugin_ec_add_cached_devices(plugin,
+									FU_DEVICE(ec_dev),
+									error))
+				return FALSE;
+		} else {
+			/* api version 2 doesn't support legacy docks */
+			if (g_error_matches(error_local, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND)) {
+				g_debug("ignoring: %s", error_local->message);
+			} else {
+				g_propagate_error(error, g_steal_pointer(&error_local));
+				return FALSE;
+			}
+		}
+		return TRUE;
+	}
+
+	if (!fu_dell_dock2_plugin_device_add(plugin, device, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static void
+fu_dell_dock2_plugin_config_mst_dev(FuPlugin *plugin)
+{
+	g_autofree const gchar *devname = NULL;
+	FuDevice *device_ec = fu_plugin_cache_lookup(plugin, "ec");
+	FuDevice *device_mst = fu_plugin_cache_lookup(plugin, "mst");
+	DellDock2EcDevType mst_devtype = DELL_DOCK2_EC_DEV_TYPE_MST;
+	DellDock2EcDevMstSubtype mst_subtype;
+
+	if (device_ec == NULL || device_mst == NULL)
+		return;
```

Comment:
does this happen regularly?  I wonder if `g_debug()` would be useful here.

Reply:
The plugin might be dealing with the other components (e.g.: rts usbhub) and at a time the `mst` is already added but no `ec`, or vice versa. Wait until the undergoing device is either a `mst` or a `ec` to proceed.

The setting should be done only once so next coming here (i.e.: both devices already added) we can leave early according to the added flag `FU_DEVICE_PRIVATE_FLAG_EXPLICIT_ORDER`.